### DISCUSSION TITLE
How to re-order atoms based on the provided upfront SMILES string

### DISCUSSION THREAD

**Discussion by dln22** [rdkit/rdkit#4179] [Q&A] [answered]
I am working on an application where we read molecules from 3d coordinates files (e.g. xyz, mol,..) for further processing. The problem is, these 3d files can have an arbitrary order of atoms which causes problems in our processing code. I have recently discovered "Inchified SMILES" that are supported by e.g. OpenBabel. I wonder if one can use such "Inchified SMILES" strings to reorder the atoms (canonicalise) of molecules read from the 3d coordinates files? If that is not possible, is there a way to reorder the atoms using SMILES strings computed by rdkit? The idea is to always end up with the same atom numbering, regardless of what was in the mol or xyz file.

Here is a pseudo-Python code that I have in mind:

```Python
mol_from_mol = rdkit.Chem.MolFromMolBlock(molBlock)          # create the molecule from the mol file
mol_from_smiI = rdkit.Chem.MolFromSmiles('Cc1cccc(c1O)O')    # create the same molecule from the provided Inchified SMILES

#Use the atoms order in mol_from_smiI  to canonicalise atoms in mol_from_mol 
```

Many thanks,
Daniel

**Comment by ptosco** [comment, marked as answer]
@dln22 `Chem.RenumberAtoms()` can do what you need:
```
import rdkit
from rdkit import Chem
from rdkit.Chem.Draw import IPythonConsole

IPythonConsole.drawOptions.addAtomIndices = True

mol_from_mol = Chem.MolFromMolBlock("""
     RDKit          2D

  9  9  0  0  0  0  0  0  0  0999 V2000
    2.5981   -0.8333    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.2990   -0.0833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.0000   -0.8333    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.0000   -2.3333    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -1.2990   -0.0833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.5981   -0.8333    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.2990    1.4167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000    2.1667    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2990    1.4167    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  2  3  2  0
  3  4  1  0
  3  5  1  0
  5  6  1  0
  5  7  2  0
  7  8  1  0
  8  9  2  0
  9  2  1  0
M  END
""")          # create the molecule from the mol file
mol_from_smiI = Chem.MolFromSmiles('Cc1cccc(c1O)O')    # create the same molecule from the provided Inchified SMILES

assert(mol_from_smiI.GetNumAtoms() == mol_from_mol.GetNumAtoms())
match = mol_from_mol.GetSubstructMatch(mol_from_smiI)
assert(match)
assert(len(match) == mol_from_mol.GetNumAtoms())
match
(5, 4, 6, 7, 8, 1, 2, 3, 0)

Chem.RenumberAtoms(mol_from_mol, match)
```
![image](https://user-images.githubusercontent.com/5244385/119564120-b42dfc00-bda8-11eb-8be5-7c7311b03e5a.png)

**Reply by dln22** [reply]
Many thanks Paolo for your answer. May I just add a quick side-question, please? Since most of our molecule files are in xyz format, we need to somehow convert them into rdkit molecule format. What we wish to do is to load them to rdkit in a way that has the same order of the atoms in xyz (including hydrogens) and then canonicalise the molecule (e.g. using the solution you posted above) and having obtained a new order, apply it to the order in xyz file. This would allow us to re-order the atoms in all of our xyz files in one, well-defined way (does this make sense?). I could see that rdkit does not have the MolFromXYZ function. I wonder if xyz -> mol ( MDL MOL format) conversion e.g. via OpenBabel can be used to load our molecules to rdkit using MolFromMol function. Do you know if the xyz->mol conversion preserves the order of the atoms  (including hydrogens)? I have also noticed that we could easily modify your solution to use inchi format as the atoms canonicalisation algorithm. We simply do not know how to refer back the obtained order to the xyz file (where hydrogens are also mapped). Once again many thanks!

**Comment by ptosco** [comment]
@dln22 you may have a look at this project from Jan Halborg Jensen's group.
https://github.com/jensengroup/xyz2mol

and also at this Jupyter Notebook of mine
https://gist.github.com/ptosco/4844d3635cf14d11e5e14381993915c1

for some inspiration.

**Reply by dln22** [reply]
Thanks for your helpful examples Paolo. Here is a function I wrote for reordering the atoms in xyz file based on the inchi atoms order:

```python
def xyzToinchiOrder(xyzFile):
    # use open babel to convert xyz to pdb then create mol_from_pdb molecule
    mol_pdb_block = obabelConv("-i","xyz",xyzFile,"-o","pdb")
    mol_from_pdb = rdkit.Chem.MolFromPDBBlock(mol_pdb_block, removeHs=False)
    
    # use open babel to generate inchi string then create mol_from_inchi molecule
    inchi = obabelConv("-i","xyz",xyzFile,"-o","inchi")
    mol_from_inchi = rdkit.Chem.inchi.MolFromInchi(inchi)
    mol_from_inchi = rdkit.Chem.rdmolops.AddHs(mol_from_inchi)

    # get the atoms order w.r.t. the inchi order
    match = mol_from_pdb.GetSubstructMatch(mol_from_inchi)

    # read the xyz file block in and reorder the atoms
    xyz = loadFromFile(xyzFile)
    xyz =xyz.split('\n')
    xyz[2:len(match)+2] = [xyz[i+2] for i in match]
    return xyz
```

I could see that in your code example you care a lot about hydrogens, bond orders and conformers. I wonder if all these extra steps are necessary for inchi atoms order conversion? Many thanks and best regards!
